src/lib/sticky-header/sticky-header-dir.ts

```diff
@@ -0,0 +1,409 @@
+      this.originalCss.width);
+    extendObject(this.element.style, stickyCss);
+  }
+
+  /**
+   * Unstuck element: When an element reaches the bottom of its cdkStickyRegion,
+   * It should be unstuck. And its position will be set as 'relative', its bottom
+   * will be set as '0'. So it will be stick at the bottom of its cdkStickyRegion and
+   * will be scrolled up with its cdkStickyRegion element. In this way, the sticky header
+   * can be changed smoothly when two sticky header meet and the later one need to replace
+   * the former one.
+   */
+  unstuckElement(): void {
+    this.isStuck = false;
+
+    if(this.stickyParent == null) {
+      return;
+    }
+
+    this.element.classList.add(STICK_END_CLASS);
+    this.stickyParent.style.position = 'relative';
+    let unstuckCss: any = this.generateCssStyle(this.originalCss.zIndex,
+      'absolute', 'auto', '0', 'auto', '0', this.originalCss.width);
+    extendObject(this.element.style, unstuckCss);
+  }
+
+
+  /**
+   * 'sticker()' function contains the main logic of sticky-header. It decides when
+   * a header should be stick and when should it be unstuck. It will first get
+   * the offsetTop of the upper scrollable container. And then get the Start and End
+   * of the sticky-header's stickyRegion.
+   * The header will be stick if 'stickyRegion Start < container offsetTop < stickyRegion End'.
+   * And when 'stickyRegion End < container offsetTop', the header will be unstuck. It will be
+   * stick to the bottom of its stickyRegion container and being scrolled up with its stickyRegion
+   * container.
+   * When 'stickyRegion Start > container offsetTop', which means the header come back to the
+   * middle of the scrollable container, the header will be reset to its
+   * original CSS.
+   * A flag, isStuck. is used in this function. When a header is stick, isStuck = true.
+   * And when the 'isStuck' flag is TRUE, the sticky-header will not be repaint, which
+   * decreases the times on repainting sticky-header.
+   */
+  sticker(): void {
+    let currentPosition: number = this.upperScrollableContainer.offsetTop;
+
+    // unstuck when the element is scrolled out of the sticky region
+    if (this.isStuck &&
+      (currentPosition < this._containerStart || currentPosition > this._scrollFinish) ||
+      currentPosition >= this._scrollFinish) {
+      this.resetElement();
+      if (currentPosition >= this._scrollFinish) {
+        this.unstuckElement();
+      }
+      this.isStuck = false;    // stick when the element is within the sticky region
+    } else if ( this.isStuck === false &&
+      currentPosition > this._containerStart && currentPosition < this._scrollFinish) {
+      this.stickElement();
+    }
+  }
+
+  defineRestrictionsAndStick(): void {
+    this.defineRestrictions();
+    this.sticker();
+  }
+
+  /**
+   * This function is used to generate a variable which contains 7 css styles.
+   * @param zIndex
```

Comment:
Removed the @param and @returns and make sure the types are correct in the function signature